### PR TITLE
Improve 'no version property' feature

### PR DESCRIPTION
**Pull request by shestee**
1. When there is no version property configured and the project is being built not from tag, the plugin deducts version based on previous tag. This commit improves plugin's behaviour in such situation with increasing patch version number, taken from last tag, by 1 and adding `-SNAPSHOT` suffix. This solves #77

2. Additionally this commit fixes bug: when `version` property is not present and `tagPrefix` has no value set (`tagPrefix=`), current plugin version wrongly picks version based on tag but should pick `0.0.1`.  Also corrected unit test for behaviour when there is no config file and code is checked out on not valid tag.

### CODE REVIEW

File: src/main/java/org/shipkit/auto/version/NextVersionPicker.java [hunk 1]
```diff
@@ -30,13 +34,18 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
...
-            Matcher matcher = pattern.matcher(tag);
+            String tag = runner.run("git", "describe", "--tags").trim();
             String result;
-            if (matcher.find()) {
-                result = matcher.group().substring(config.getTagPrefix().length());
-                explainVersion(log, result, "deducted version based on tag: '" + config.getTagPrefix() + result + "'");
+            if (isVersionTag(tag, config.getTagPrefix()) && isSupportedVersion(tag.substring(config.getTagPrefix().length()))) {
+                result = tag.substring(config.getTagPrefix().length());
+                explainVersion(log, result, "deducted version based on tag: '" + tag + "'");
+            } else if (isVersionTag(tag, config.getTagPrefix()) && isSnapshot(tag, config.getTagPrefix())){
```
Comment (mockitoguy) [inline src/main/java/org/shipkit/auto/version/NextVersionPicker.java]: Let's push the logic of `isVersionTag` to `isSnapshot` method. I think it will be cleaner. It's OK to do this in a separate PR. Bonus: we might do the same with `isSupportedVersion`.

File: src/main/java/org/shipkit/auto/version/NextVersionPicker.java [hunk 2]
```diff
@@ -30,13 +34,18 @@ String pickNextVersion(Optional<Version> previousVersion, VersionConfig config,
...
+                Pattern pattern = Pattern.compile("\\d+\\.\\d+\\.\\d+");
+                Matcher matcher = pattern.matcher(tag);
+                matcher.find();
+                result = Version.valueOf(matcher.group()).incrementPatchVersion("SNAPSHOT").toString();
```
Comment (mockitoguy) [inline src/main/java/org/shipkit/auto/version/NextVersionPicker.java:45]: Nice use of the `Version` library!

File: src/main/java/org/shipkit/auto/version/VersionConfig.java [hunk 3]
```diff
@@ -99,6 +99,10 @@ static boolean isSupportedVersion(String version) {
         return version.matches("\\d+\\.\\d+\\.\\d+");
     }
 
+    static boolean isSnapshot(String tag, String tagPrefix) {
```
Comment (mockitoguy) [inline src/main/java/org/shipkit/auto/version/VersionConfig.java:112]: Let's add javadoc

**Review by mockitoguy — APPROVED** [review]
Nice, please consider my feedback!

File: src/main/java/org/shipkit/auto/version/VersionConfig.java (hunk 3 above)
Comment (shestee) [reply src/main/java/org/shipkit/auto/version/VersionConfig.java:112]: @mockitoguy Thanks for your review! As you suggested I refactored isSupportedVersion() and isSnapshot() and covered these methods with unit tests. I also added javadoc.

File: src/test/groovy/org/shipkit/auto/version/VersionConfigTest.groovy [hunk 4]
```diff
@@ -43,17 +44,32 @@ class VersionConfigTest extends TmpFolderSpecification {
...
-        '2.33.444'  | true
+        'v1.0.0'    | 'v'        | true
+        'v2.33.444' | 'v'        | true
 
-        'x'         | false
-        '1.0'       | false
-        '1.0.0-rc'  | false
+        'x'         | ''         | false
+        'v1.0'      | 'v'        | false
+        'v1.0.0-rc' | 'v'        | false
```
Comment (mockitoguy) [inline src/test/groovy/org/shipkit/auto/version/VersionConfigTest.groovy:62]: Let's add few more negative cases:
 - good version without tag prefix: `1.2.3`
 - good version with different tag prefix: `ver-1.2.3`

**Review by mockitoguy — APPROVED** [review]
Let's tighten the unit test and ship it!

File: src/test/groovy/org/shipkit/auto/version/VersionConfigTest.groovy (hunk 4 above)
Comment (shestee) [reply src/test/groovy/org/shipkit/auto/version/VersionConfigTest.groovy:62]: Good point! I have just added some more cases.
